### PR TITLE
support scoped package and using global npm

### LINKED ISSUE #78: scoped packages appear to be unsupported
[scoped packages](https://docs.npmjs.com/misc/scope) doesn't appear to be supported:

snippet from `package.json`:
```json
"@heroku/worker-monitor": "^1.0.0"
```

output from `npm-shrinkwrap`:
```
Problems were encountered
Please correct and try again.
missing: @heroku/worker-monitor@^1.0.0, required by dashboard-server@
something went wrong. Did not write npm-shrinkwrap.json
```

### PR DESCRIPTION
**Pull request by th507**
- support scoped package (Fix #78, must run w/ npm >=2.0, see `--use-global-npm`)
- support calling global instead of local/bundled npm (via `--use-global-npm`)

### CODE REVIEW

**Comment by th507** [comment]
I forgot to add test case for scoped package support. Will do soon.

**Comment by th507** [comment]
Test case added.

**Comment by djanowski** [comment]
I'm testing this branch and I get this error:

```
$ npm-shrinkwrap --use-global-npm
Problems were encountered
Please correct and try again.
invalid: readable-stream@1.0.27-1 /path/to/app/node_modules/npm-shrinkwrap/node_modules/npm/node_modules/sha/node_modules/readable-stream
invalid: string_decoder@0.10.25-1 /path/to/app/node_modules/npm-shrinkwrap/node_modules/npm/node_modules/sha/node_modules/readable-stream/node_modules/string_decoder
something went wrong. Did not write npm-shrinkwrap.json
```

Any ideas?

**Comment by Raynos** [comment]
@djanowski that's a classic npm1 vs npm2 issue.

File: package.json [hunk 1]
```diff
@@ -31,12 +31,16 @@
     "fixtures-fs": "^2.0.0",
     "istanbul": "~0.3.2",
     "jshint": "2.5.6",
+    "mkdirp": "^0.5.1",
     "pre-commit": "0.0.9",
     "tap-spec": "~1.0.0",
     "tape": "^3.0.2"
   },
   "scripts": {
+    "pretest": "rm -fr test/proj /tmp/proj",
```
Comment (Raynos) [inline package.json]: please dont use pre scripts.

The correct way is to add a `rimraf('./proj'); rimraf('/tmp/proj')` to the source code.

File: sync/purge-excess.js [hunk 2]
```diff
@@ -39,39 +39,54 @@ function purgeExcess(dir, shrinkwrap, opts, cb) {
...
-        });
-
-        if (opts.dev && shrinkwrap.devDependencies) {
-            var devDeps = shrinkwrap.devDependencies;
-            var devKeys = Object.keys(devDeps).map(function (s) {
-                return s.toLowerCase();
-            });
+    var files = [];
+    try {
+      files = fs.readdirSync(dir);
```
Comment (Raynos) [inline sync/purge-excess.js]: sync IO is banned. please async.

File: test/git-https-use-case.js [hunk 3]
```diff
@@ -14,6 +14,9 @@ test('npm-shrinkwrap --dev on git+https uri', fixtures(__dirname, {
         'package.json': JSON.stringify({
             name: 'foo',
             version: '0.1.0',
+            description: "",
```
Comment (Raynos) [inline test/git-https-use-case.js]: are these fields necessary ?

File: test/npm-shrinkwrap.js [hunk 4]
```diff
@@ -2,12 +2,27 @@ var test = require('tape');
 var fixtures = require('fixtures-fs');
 var path = require('path');
 var fs = require('fs');
+var mkdirp = require("mkdirp");
+
+// a hack to workaround fixtures-fs
+// override fs.mkdirp and leaving all other functions intact
+var _fs = Object.create(fs);
+_fs.mkdir = mkdirp;
```
Comment (Raynos) [inline test/npm-shrinkwrap.js]: why ?

File: test/npm-shrinkwrap.js [hunk 5]
```diff
@@ -204,14 +272,21 @@ test('error on additional GIT module', fixtures(__dirname, {
...
+        'node_modules': {},
+        nonono: 1
     })
 }, function (assert) {
-    npmShrinkwrap(PROJ, function (err) {
+    npmShrinkwrap(OPT, function (err) {
         assert.ok(err);
 
         assert.notEqual(err.message.indexOf(
-            'missing: foo@git://github.com:uber/foo#v1.0.0'), -1);
```
Comment (Raynos) [inline test/npm-shrinkwrap.js]: why this churn ?

File: test/npm-shrinkwrap.js [hunk 6]
```diff
@@ -204,14 +272,21 @@ test('error on additional GIT module', fixtures(__dirname, {
         dependencies: {
             'foo': 'git://github.com:uber/foo#v1.0.0'
         },
-        'node_modules': {}
+        'node_modules': {},
+        nonono: 1
```
Comment (Raynos) [inline test/npm-shrinkwrap.js]: what is this nonono ?

**Comment by Raynos** [comment]
It looks like you only changed the `sync` logic and not any of the other logic.

I suspected that there is a bunch of broken regex that cant match `@foo/bar` instead of `bar`

**Comment by djanowski** [comment]
@Raynos I'm on npm v2.11. I assumed this PR adds the option to use global npm as bundled npm is 1.x?

File: test/npm-shrinkwrap.js (hunk 6 above)
Comment (th507) [inline test/npm-shrinkwrap.js]: My bad. Fixed

File: test/npm-shrinkwrap.js (hunk 5 above)
Comment (th507) [inline test/npm-shrinkwrap.js]: On my machine (nvm + iojs v2.3.0 + npm 2.11.1) `err.meesage` is `foo@git://github.com/%3Auber/foo.git#v1.0.0` , which breaks this assertion.
It adds a `.git` after package name.

My guess (could be wrong) is that npm 2+ might generate a slightly different error description.

File: test/git-https-use-case.js (hunk 3 above)
Comment (th507) [inline test/git-https-use-case.js]: I add these to avoid nom warning messages pop up during tests, warning like this:
```
npm WARN package.json foo@0.1.0 No description
npm WARN package.json foo@0.1.0 No repository field.
npm WARN package.json foo@0.1.0 No README data
```

If that's not a problem, I could revert these changes.

File: sync/purge-excess.js (hunk 2 above)
Comment (th507) [inline sync/purge-excess.js]: Is there a good argument to avoid sync IO altogether in this package?

IMHO, `npm-shrinkwrap` is mainly used as a CLI command and sync/async IO is not all that different in this scenario. Or is it something else I missed?

File: test/npm-shrinkwrap.js (hunk 4 above)
Comment (th507) [inline test/npm-shrinkwrap.js]: `mkdirp` is used to create folders like `@th507/foo` in one stroke.
 
In `fixtures-fs` package, `create-fixtures.js` line 28, it uses `opts.fs.mkdir` instead of `opts.mkdirp` to make directory.

When I wish to override `mkdir` behavior, I have to provide a fake `opts.fs`, not `opts.mkdirp`. 
I tried `opts.fs = { mkdir: function(){...} }`, but it failed because the script required other function/property from `opts.fs` object.

I did tried to patch `create-fixtures.js`. However, simply replacing `fs.mkdir` with `mkdirp` broke one of the unit tests, and there was not simpler solution in sight.

I could work on `fixtures-fs` package later and then remove the hack.
But I think maybe we could focus on bringing (basic) scoped package support `npm-shrinkwrap` for the moment.

**Comment by th507** [comment]
@Raynos Could you tell where those possible failing regex are? I could take a quick look at them.

File: sync/purge-excess.js (hunk 2 above)
Comment (Raynos) [inline sync/purge-excess.js]: because the `sync` command does fanout over the disk; it will be a lot slower synchronously.

Writing async logic is not hard; I generally ban sync logic.

File: test/git-https-use-case.js (hunk 3 above)
Comment (Raynos) [inline test/git-https-use-case.js]: looks fine.

File: test/npm-shrinkwrap.js (hunk 4 above)
Comment (Raynos) [inline test/npm-shrinkwrap.js]: @th507 

Then just do

```js
{
    'node_modules': {
        'th507': {
            'foo': moduleFixture(...)
        }
    }
}
```

File: test/npm-shrinkwrap.js (hunk 5 above)
Comment (Raynos) [inline test/npm-shrinkwrap.js]: This is fine.

File: test/npm-shrinkwrap.js (hunk 4 above)
Comment (th507) [inline test/npm-shrinkwrap.js]: Good point. Fixed.

File: sync/purge-excess.js (hunk 2 above)
Comment (th507) [inline sync/purge-excess.js]: Switched to async API as requested.

**Comment by th507** [comment]
Any thoughts?

**Comment by th507** [comment]
Hi, @djanowski @Raynos.
Is there anything I could do?

**Comment by Raynos** [comment]
Looks legit. Let me try this branch on an internal project of ours then i'll merge.

**Comment by th507** [comment]
Thanks!

**Comment by jessepedler** [comment]
Any update on this?  Were you able to test it on an internal project?  We'd love to use `npm-shrinkwrap` and dig ourselves out of shrinkwrap hell, but we make heavy use of scoped packages.

**Comment by th507** [comment]
I've updated my patch to respect `opts.useGlobalNPM` config in `sync/index.js`.

**Comment by geertweening** [comment]
Any update on this? We started using scoped packages which has been problematic with `npm-shrinkwrap`

**Comment by sdepold** [comment]
I'd like to have this too

**Comment by nathanbrauer** [comment]
:+1:

**Comment by joanniclaborde** [comment]
:+1:

**Comment by doughsay** [comment]
+1

**Comment by doughsay** [comment]
I tried running this branch against a project with private packages, still doesn't work.  I got something like this:

```
~/lib/npm-shrinkwrap/bin/cli.js
Problems were encountered
Please correct and try again.
missing: <private package>
extraneous: <public dependency of private package>
extraneous: <public dependency of private package>
...
something went wrong. Did not write npm-shrinkwrap.json
```

npm 3.6.0
node 5.6.0

**Comment by ljharb** [comment]
Could perhaps these two features be split out into separate PRs? It also looks like perhaps they need updating and rebasing.

**Comment by mattfysh** [comment]
@Raynos what is the status of this pull request? Can it be merged, or is it too out of date? Happy to submit a new PR branched from latest master

**Comment by lili21** [comment]
Any update ?

**Comment by mking-clari** [comment]
We cannot use this tool without scoped packages. Most of our code is private and thus uses scoped npm packages (in our case, hosted in [Sinopia](https://github.com/rlidwka/sinopia) but applies equally to npm's private registry solution).

**Comment by mlmorg** [comment]
Closing since #114 fixed this.
